src/backend/distributed/planner/distributed_planner.c

```diff
@@ -550,30 +544,111 @@ IsModifyDistributedPlan(DistributedPlan *distributedPlan)
 }
 
 
+/*
+ * FastPathDistributedStmt creates a distributed planned statement using the
+ * FastPathPlanner.
+ */
+static PlannedStmt *
+FastPastDistributedStmt(DistributedStmtOptions *opts, Const *distributionKeyValue)
+{
+	opts->plannerRestrictionContext->fastPathRestrictionContext->fastPathRouterQuery =
+		true;
+	opts->plannerRestrictionContext->fastPathRestrictionContext->distributionKeyValue =
+		distributionKeyValue;
+
+	PlannedStmt *fastPathPlan = FastPathPlanner(opts->originalQuery, opts->parse,
+												opts->boundParams);
+
+	return CreateDistributedPlannedStmt(fastPathPlan, opts);
+}
+
+
+/*
+ * NonFastPathDistributedStmt creates a distributed planned statement using the
+ * PG planner.
+ */
+static PlannedStmt *
+NonFastPathDistributedStmt(DistributedStmtOptions *opts,
+						   List *rangeTableList,
+						   int rteIdCounter)
+{
+	/*
+	 * Call into standard_planner because the Citus planner relies on both the
+	 * restriction information per table and parse tree transformations made by
+	 * postgres' planner.
+	 */
+	PlannedStmt *standardPlan = standard_planner(opts->parse, opts->cursorOptions,
+												 opts->boundParams);
+
+	/* may've inlined new relation rtes */
+	rangeTableList = ExtractRangeTableEntryList(opts->parse);
+	rteIdCounter = AssignRTEIdentities(rangeTableList, rteIdCounter);
+
+
+	PlannedStmt *result = CreateDistributedPlannedStmt(standardPlan, opts);
+
+	bool setPartitionedTablesInherited = true;
+	AdjustPartitioningForDistributedPlanning(rangeTableList,
+											 setPartitionedTablesInherited);
+
+	return result;
+}
+
+
+/*
+ * NonDistributedStmt creates a normal (non-distributed) planned statement
+ * using the PG planner.
+ */
+static PlannedStmt *
+NonDistributedStmt(DistributedStmtOptions *opts)
+{
+	PlannedStmt *result = standard_planner(opts->parse, opts->cursorOptions,
+										   opts->boundParams);
+
+	bool hasExternParam = false;
+	DistributedPlan *delegatePlan = TryToDelegateFunctionCall(opts->parse,
+															  &hasExternParam);
```

Comment:
Would it make sense to capture function call delegation as a separate `else if` in `distributed_planner`? 

e.g.

```c
		if (fastPathRouterQuery)
		{
			result = FastPastDistributedStmt(&opts, distributionKeyValue);
		}
		else if (needsDistributedPlanning)
		{
			result = NonFastPathDistributedStmt(&opts, rangeTableList, rteIdCounter);
		}
		else if (TryToDelegateFunctionCall(opts->parse, &hasExternParams, &delegatePlan))
		{
			result = PlanFunctionCallDelegation(result, hasExternParams, delegatePlan);
		}
		else
		{
			result = standard_planner(opts.parse, opts.cursorOptions, opts.boundParams);
		}
```